File: hands/remote/playout.go
package main

import (
	"strings"

	"github.com/pion/interceptor"
	"github.com/pion/rtp"
	"github.com/pion/webrtc/v4"
)

const screenPlayoutDelayURI = "http://www.webrtc.org/experiments/rtp-hdrext/playout-delay"

// Screen receivers should render promptly instead of accumulating frames for
// smooth movie playback. Request a 0–100ms best-effort range, leaving
// room for jitter and audio synchronization at the receiver.
// Negotiate the extension: older viewers continue using their normal behavior.
func screenPeerAPI(settings webrtc.SettingEngine) (*webrtc.API, error) {
	media := &webrtc.MediaEngine{}
	if err := media.RegisterDefaultCodecs(); err != nil {
		return nil, err
	}
	if err := media.RegisterHeaderExtension(webrtc.RTPHeaderExtensionCapability{URI: screenPlayoutDelayURI}, webrtc.RTPCodecTypeVideo); err != nil {
		return nil, err
	}
	registry := &interceptor.Registry{}
	// Preserve Pion's NACK, RTCP reports and other default interceptors.
	if err := webrtc.RegisterDefaultInterceptors(media, registry); err != nil {
		return nil, err
	}
	registry.Add(screenPlayoutFactory{})
	return webrtc.NewAPI(webrtc.WithSettingEngine(settings), webrtc.WithMediaEngine(media), webrtc.WithInterceptorRegistry(registry)), nil
}

type screenPlayoutFactory struct{}

func (screenPlayoutFactory) NewInterceptor(string) (interceptor.Interceptor, error) {
	return &screenPlayoutInterceptor{}, nil
}

type screenPlayoutInterceptor struct{ interceptor.NoOp }

func (*screenPlayoutInterceptor) BindLocalStream(info *interceptor.StreamInfo, writer interceptor.RTPWriter) interceptor.RTPWriter {
	if !strings.HasPrefix(strings.ToLower(info.MimeType), "video/") {
		return writer
	}
	for _, extension := range info.RTPHeaderExtensions {
		if extension.URI != screenPlayoutDelayURI || extension.ID < 1 || extension.ID > 255 {
			continue
		}
		id := uint8(extension.ID)
		// Wire units are 10ms. Repeat on every packet so joins, loss and ICE
		// restarts need no extra acknowledgement or per-viewer state.
		delay, _ := (rtp.PlayoutDelayExtension{MinDelay: 0, MaxDelay: 10}).Marshal()
		return interceptor.RTPWriterFunc(func(header *rtp.Header, payload []byte, attributes interceptor.Attributes) (int, error) {
			// StaticRTP shares headers across viewers. Never leak this viewer's
			// negotiated extension ID into another viewer's packet.
			out := header.Clone()
			if err := out.SetExtension(id, delay); err != nil {
				return 0, err
			}
			return writer.Write(&out, payload, attributes)
		})
	}
	return writer
}

File: crates/nanocodex-remote/src/playout.rs
//! Negotiated screen playout hint, matching the Go Hand's 0–100ms range.
//! Keep this per stream: a shared track can have different extension IDs for
//! each viewer, including viewers that do not negotiate the extension at all.
use async_trait::async_trait;
use std::sync::Arc;
use webrtc::{
    api::media_engine::MediaEngine,
    interceptor::{
        Attributes, Error, Interceptor, InterceptorBuilder, RTCPReader, RTCPWriter, RTPReader,
        RTPWriter, registry::Registry, stream_info::StreamInfo,
    },
    rtp::{
        header::{EXTENSION_PROFILE_ONE_BYTE, EXTENSION_PROFILE_TWO_BYTE, Extension},
        packet::Packet,
    },
    rtp_transceiver::rtp_codec::{RTCRtpHeaderExtensionCapability, RTPCodecType},
};

const URI: &str = "http://www.webrtc.org/experiments/rtp-hdrext/playout-delay";
type InterceptorResult<T> = std::result::Result<T, Error>;

pub(crate) fn register(
    engine: &mut MediaEngine,
    mut registry: Registry,
) -> crate::Result<Registry> {
    engine.register_header_extension(
        RTCRtpHeaderExtensionCapability { uri: URI.into() },
        RTPCodecType::Video,
        None,
    )?;
    // Extend the caller's default NACK, report, and transport feedback chain.
    registry.add(Box::new(PlayoutDelay));
    Ok(registry)
}

struct PlayoutDelay;
impl InterceptorBuilder for PlayoutDelay {
    fn build(&self, _id: &str) -> InterceptorResult<Arc<dyn Interceptor + Send + Sync>> {
        Ok(Arc::new(Self))
    }
}

#[async_trait]
impl Interceptor for PlayoutDelay {
    async fn bind_local_stream(
        &self,
        info: &StreamInfo,
        writer: Arc<dyn RTPWriter + Send + Sync>,
    ) -> Arc<dyn RTPWriter + Send + Sync> {
        if !info
            .mime_type
            .split_once('/')
            .is_some_and(|(kind, _)| kind.eq_ignore_ascii_case("video"))
        {
            return writer;
        }
        match info
            .rtp_header_extensions
            .iter()
            .find(|ext| ext.uri == URI && (1..=255).contains(&ext.id))
        {
            Some(extension) => Arc::new(PlayoutWriter {
                writer,
                // Two 12-bit fields, in 10ms units: minimum 0, maximum 10.
                // Repeat on every packet for joins, packet loss, and ICE restarts.
                extension: Extension {
                    id: extension.id as u8,
                    payload: vec![0, 0, 10].into(),
                },
            }),
            None => writer,
        }
    }

    async fn bind_rtcp_reader(
        &self,
        reader: Arc<dyn RTCPReader + Send + Sync>,
    ) -> Arc<dyn RTCPReader + Send + Sync> {
        reader
    }
    async fn bind_rtcp_writer(
        &self,
        writer: Arc<dyn RTCPWriter + Send + Sync>,
    ) -> Arc<dyn RTCPWriter + Send + Sync> {
        writer
    }
    async fn unbind_local_stream(&self, _info: &StreamInfo) {}
    async fn bind_remote_stream(
        &self,
        _info: &StreamInfo,
        reader: Arc<dyn RTPReader + Send + Sync>,
    ) -> Arc<dyn RTPReader + Send + Sync> {
        reader
    }
    async fn unbind_remote_stream(&self, _info: &StreamInfo) {}
    async fn close(&self) -> InterceptorResult<()> {
        Ok(())
    }
}

struct PlayoutWriter {
    writer: Arc<dyn RTPWriter + Send + Sync>,
    extension: Extension,
}
#[async_trait]
impl RTPWriter for PlayoutWriter {
    async fn write(&self, packet: &Packet, attributes: &Attributes) -> InterceptorResult<usize> {
        // Never mutate the shared track packet or another viewer's extensions.
        let mut outgoing = packet.clone();
        // The RTP helper chooses its header format from payload length alone.
        // IDs above 14 require the two-byte format even for this 3-byte hint.
        // Promote existing one-byte extensions and recalculate their padding.
        if self.extension.id > 14
            && (!outgoing.header.extension
                || outgoing.header.extension_profile == EXTENSION_PROFILE_ONE_BYTE)
        {
            outgoing.header.extension = true;
            outgoing.header.extension_profile = EXTENSION_PROFILE_TWO_BYTE;
            let size: usize = outgoing
                .header
                .extensions
                .iter()
                .map(|ext| ext.payload.len() + 2)
                .sum();
            outgoing.header.extensions_padding = (4 - size % 4) % 4;
        }
        if outgoing
            .header
            .set_extension(self.extension.id, self.extension.payload.clone())
            .is_err()
        {
            // An optional playback preference must not interrupt media using an
            // unsupported header format. Preserve the original packet instead.
            return self.writer.write(packet, attributes).await;
        }
        self.writer.write(&outgoing, attributes).await
    }
}

#[cfg(test)]
#[path = "playout_tests.rs"]
mod tests;
